hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -3274,14 +3395,27 @@ def construct_expr(ir: IR,
         return Expression(ir, None, indices, aggregations)
     if isinstance(type, tarray) and is_numeric(type.element_type):
         return ArrayNumericExpression(ir, type, indices, aggregations)
+    if isinstance(type, tarray):
+        etype = type.element_type
+        if isinstance(etype, (hl.tarray, hl.tset)):
+            while isinstance(etype, (hl.tarray, hl.tset)):
+                etype = etype.element_type
+        if isinstance(etype, hl.tstruct):
+            return ArrayStructExpression(ir, type, indices, aggregations)
+    if isinstance(type, tset) and isinstance(type.element_type, tstruct):
+        etype = type.element_type
+        if isinstance(etype, (hl.tarray, hl.tset)):
+            while isinstance(etype, (hl.tarray, hl.tset)):
+                etype = etype.element_type
+        if isinstance(etype, hl.tstruct):
+            return SetStructExpression(ir, type, indices, aggregations)
     if isinstance(type, tndarray) and is_numeric(type.element_type):
         return NDArrayNumericExpression(ir, type, indices, aggregations)
-    elif type in scalars:
+    if type in scalars:
```

Comment:
pylint doesn't like the `else` `elif`s if not necessary. at first I hated it but now I prefer it.

Reply:
do you have a PEP reference?

Reply:
It's not in PEP-8, which quietly demonstrates both styles in an unrelated recommendation. In pylint it's called `no-else-return`. We can add this to our standard pylint rules if you prefer this style. At first I hated it. Now I've come to prefer the uniformity of `if` and the lack of indentation on the else case.

Reply:
I suppose if all of hail is written using the else-return style we should stay consistent.

Reply:
For what it's worth, I think standard C++ style is also to never put an else clause on an if statement that returns.

Reply:
I'm happy with either, but do have a preference for `elif` in this case. I'm fine with going with style guidelines, though.

I think I have somewhat mixed preferences, too. I'd prefer:

```python
def foo(*args):
    if take_other_path:
        return blah(*args)
    ...
    lots of code
    ...
    return other_thing
```

to

```python
def foo(*args):
    if take_other_path:
        return blah(*args)
    else:
        lots of code
        ...
        return other_thing
```

However, in code with several short conditionals, I prefer 

```python
def foo(a, b, c, ..., z):
    if a < 0:
        return a
    elif b < 0:
        return b
    ...
    else:
        return z
```

to 

```python
def foo(a, b, c, ..., z):
    if a < 0:
        return a
    if b < 0:
        return b
    ...
    return z
```

, since if you forget one of the "return" keywords, you get a piece of code that's much harder to debug.